### ISSUE TITLE
SimplifyBooleanReturn false negatives

### ISSUE THREAD

**Issue by jungm** [checkstyle/checkstyle#11608]
I have read check documentation: https://checkstyle.org/config_coding.html#SimplifyBooleanReturn
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <module name="TreeWalker">
    <module name="SimplifyBooleanReturn"/>
  </module>
</module>

/var/tmp $ cat Test.java
// Taken from tests: https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanreturn/InputSimplifyBooleanReturn.java
public class Test
{
    public static boolean isOddMillis()
    {
        boolean even = System.currentTimeMillis() % 2 == 0;

        // can be simplified to "if (even)"
        if (even == true) { // violation
            return false;
        }
        else {
            return true;
        }
        // return can be simplified to "return !even"
    }

    public static boolean isOddMillis2()
    {
        boolean even = System.currentTimeMillis() % 2 == 0;
        // can be simplified to "return !even"
        if (!even) // violation
            return true;
        else
            return false;
    }

    public static boolean giveMeTrue()
    {
        boolean tt = isOddMillis() || true;
        boolean ff = isOddMillis() && false;
        return !false || (true != false);
    }

    public boolean ifNoElse()
    {
        if (isOddMillis()) {
            return true;
        }
        return false;
    }

    boolean a() {
        boolean asd = false;
        boolean dasa = true;

        if (asd) {
            return true;
        } else {
            return dasa;
        }
    }

    boolean aa() {
        boolean asd = false;
        boolean dasa = true;

        if (asd) {
            return dasa;
        } else {
            return true;
        }
    }

    boolean b() {
        boolean asd = false;

        if(asd);
        else;

        return true;
    }
}


/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java -jar $RUN_LOCALE -jar checkstyle-10.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/checkstyle/Test.java:9:9: Conditional logic can be removed. [SimplifyBooleanReturn]
[ERROR] /var/tmp/checkstyle/Test.java:22:9: Conditional logic can be removed. [SimplifyBooleanReturn]
Audit done.
Checkstyle ends with 2 errors.
```

---

I expect this check to also handle other cases from https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/simplifybooleanreturn/InputSimplifyBooleanReturn.java, though it seems like only the first two cases are currently being handled.

**Comment by strkkk** [comment]
@jungm can you point to specific lines where you expect a violation?

Please also note that checkstyle cannot resolve types.

**Comment by jungm** [comment]
@strkkk The case I specifically encountered was like the one in `ifNoElse`, so the `if` in line 37 should be a violation.  I'd also expect to see violations in `a` (line 47) and `aa` (line 58), but without resolving the types of those variables and detecting that they're constant this will not be easily possible.

**Comment by jungm** [comment]
@romani, is this something I can pick up? It's an old issue of mine
I can still reproduce this in the latest checkstyle release

**Comment by romani** [comment]
It is not approved yet, let me review it now, it might be just lost in storm of issues and PRs.

**Comment by romani** [comment]
>`if (isOddMillis()) { // ACTUAL ISSUE - violation expected but none reported`

In specially this case we can  detect that this method is Boolean in return.
But checkstyle is not type aware tool https://checkstyle.org/writingchecks.html#Limitations so as method start to get parameters, it becomes very complicated to validate and in most cases prone to false positives.
>
```
if (boolVar) { // ACTUAL ISSUE - violation expected but none reported
            return true;
        }
        return false;
```

This case we can reliably to catch.

Beware that we prefer false negatives in comparison to false positives, so we often skip validation for cases where we can not reliably report violation.

To approve this issue I need to clarify scope or target of issue to more specific use case.
We can split issue into two, to fix simple problems first, and deal with complicated or controversial separately.

Please update issue description.

**Comment by romani** [comment]
There are few more cases under `// ACTUAL ISSUE - violation expected` do you expect violations on them also?

Please update issue description with trailing comments to point to line where you expect violation

**Comment by jungm** [comment]
Initially I did expect those below the `ACTUAL ISSUE` comment as well to fail, wasn't aware of checkstyles scope/limitations when I initially created the issue to be honest

I've clarified more with comments what I do still think should fail and whats okay for me to pass. Only this snippet is left now, I still do think that might be a bug:
```java
       if (boolVar) { // ACTUAL ISSUE - violation expected but none reported
            return true;
        }
        return false;
```

**Comment by rnveach** [comment]
@romani 
> But checkstyle is not type aware tool... it becomes very complicated to validate and in most cases prone to false positives.

I am not seeing anything in this needing type aware. The same code is already supported if we use else instead of a drop through (see below). This check does not look at the contents of the if, and we are safe assuming the expression of the if always equates to a boolean since that is what an if statement does.

````
$ cat TestClass.java
public class TestClass {
    void method1() {
        if (isOddMillis()) {
            return true;
        }
        return false;
    }

    void method2() {
        if (isOddMillis()) {
            return true;
        } else {
            return false;
        }
    }

    void method3() {
        if (isOddMillis())
            return true;
        else
            return false;
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <property name="haltOnException" value="false"/>

    <module name="TreeWalker">
    <module name="SimplifyBooleanReturn"/>
    </module>
</module>

$ java -jar checkstyle-10.12.4-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:10:9: Conditional logic can be removed. [SimplifyBooleanReturn]
[WARN] TestClass.java:18:9: Conditional logic can be removed. [SimplifyBooleanReturn]
Audit done.
````

**Comment by romani** [comment]
Yes @rnveach is right, we can rely on compiler, if something is in IF expression, it returns boolean.

Let's make issue a bit more explicit and short.
Let's keep input file as:
```
public class TestClass {
    void method1() {
        if (isOddMillis()) {
            return true;
        }
        return false;
    }
}
```

And let's vote by maintainers to extend Check to catch such cases.
I am ok to extend.

**Comment by rnveach** [comment]
I am ok to extend as long as there is no other statements before the returns.

**Comment by romani** [comment]
@jungm , please update issue description to show only this case and I am approving issue.

**Comment by jungm** [comment]
@romani issue description is updated

**Comment by romani** [comment]
@jungm , issue is approved.
Documentation in xdoc folder need to be updated too to show that it becomes new target of Check.

@nrmancuso , fyi.
